### PR TITLE
Use one time publisher

### PR DESCRIPTION
**Pull request by Sammers21**
Every storage implementation has been updated in accordance with #206 puzzle.

_Closes #206

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#207](https://github.com/artipie/asto/issues/207) is assigned to @vatavuk[/z](https://www.0crat.com/u/vatavuk), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/5de4bcb7-6510-462a-9115-c23e60b31010); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by Sammers21** [comment]
> there is no point in wrapping incoming Content in save methods, but this should be done in tests

There is a point: when we receive a publisher we want to make sure it is consumed only once by a particular storage implementation

**Comment by 0crat** [comment]
@g4s8[/z](https://www.0crat.com/u/g4s8) I see that you unassigned this issue; the order is still assigned to @vatavuk[/z](https://www.0crat.com/u/vatavuk) though; to cancel the order use `refuse`, as in [§6](http://www.zerocracy.com/policy.html#6)

**Comment by 0crat** [comment]
This pull request [#207](https://github.com/artipie/asto/issues/207) is assigned to @olegmoz[/z](https://www.0crat.com/u/olegmoz), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/41e17a2b-67bf-4043-a339-23c79cbd242a); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; @olegmoz[/z](https://www.0crat.com/u/olegmoz) is not a member of this project yet, but they can request to join, as [§1](http://www.zerocracy.com/policy.html#1) explains; there will be no monetary reward for this job

**Comment by olegmoz** [comment]
> There is a point: when we receive a publisher we want to make sure it is consumed only once by a particular storage implementation

Why we want to make sure in runtime? Should the application crash if a publisher readable more then once to the `Storage`? I doubt it.
At the same time I strongly agree that we need to check that `Storage` works with `Publisher`s readable once, but it should be checked in tests, not in runtime

**Comment by Sammers21** [comment]
> Why we want to make sure in runtime? Should the application crash if a publisher readable more then once to the Storage? I doubt it.

Yes, I think the request should fail if a publisher consumed for the second time. In full accordance with [fail-fast](https://en.wikipedia.org/wiki/Fail-fast) principle.

**Comment by olegmoz** [comment]
> Yes, I think the request should fail if a publisher consumed for the second time. In full accordance with [fail-fast](https://en.wikipedia.org/wiki/Fail-fast) principle.

For the third time, why you think you need to check it in runtime? This could be (and should be)  verified in tests

**Comment by Sammers21** [comment]
@olegmoz, if it fails in runtime, it means that it is not covered by unit tests and the bug should be fixied

**Comment by olegmoz** [comment]
@Sammers21 you do not really answer my question

**Comment by Sammers21** [comment]
@olegmoz, we want to make sure that a publisher is consumed only once in runtime because we want users to report a bug when it occurs and, If a test suit could not find a bug, it should not stay hidden. We want this bug to be discovered as soon as possible with respect to [fail-fast](https://en.wikipedia.org/wiki/Fail-fast) principle.

**Comment by olegmoz** [comment]
@Sammers21 we may check that `Publisher` is read only once by `Storage` implementation in tests, without releasing buggy software to users. That is whole point of having unit tests. If such tests are done, then there is no point in keeping the check in runtime. Fail-fast principle has nothing to do with it.

**Comment by Sammers21** [comment]
> @Sammers21 we may check that Publisher is read only once by Storage implementation in tests, without releasing buggy software to users

But what if tests were not able to a bug?

**Comment by Sammers21** [comment]
@olegmoz, review the PR once again please

File: src/test/java/com/artipie/asto/memory/InMemoryStorageTest.java [hunk 1]
```diff
@@ -55,9 +55,11 @@ void shouldNotBeBlockedByEndlessContent() throws Exception {
...
         this.storage.save(
             key,
-            new Content.From(
-                ignored -> {
-                }
+            new Content.OneTime(
+                new Content.From(
+                    ignored -> {
+                    }
+                )
```
Comment (olegmoz) [inline src/test/java/com/artipie/asto/memory/InMemoryStorageTest.java:62]: @Sammers21 this wrapping is not needed

File: src/test/java/com/artipie/asto/s3/S3StorageTest.java [hunk 2]
```diff
@@ -105,7 +105,7 @@ void shouldUploadObjectWhenSaveContentOfUnknownSize(final AmazonS3 client) throw
         final String key = "unknown/size";
         this.storage().save(
             new Key.From(key),
-            new Content.From(new OneOffPublisher(ByteBuffer.wrap(data)))
+            new Content.OneTime(new Content.From(new OneOffPublisher(ByteBuffer.wrap(data))))
```
Comment (olegmoz) [inline src/test/java/com/artipie/asto/s3/S3StorageTest.java:108]: @Sammers21 `OneOffPublisher` usage becomes pointless here and in all other places, this class may be safely removed

File: src/test/java/com/artipie/asto/s3/S3StorageTest.java [hunk 3]
```diff
@@ -119,7 +119,7 @@ void shouldUploadObjectWhenSaveLargeContent(final AmazonS3 client) throws Except
         final String key = "big/data";
         this.storage().save(
             new Key.From(key),
-            new Content.From(new OneOffPublisher(ByteBuffer.wrap(data)))
+            new Content.OneTime(new Content.From(new OneOffPublisher(ByteBuffer.wrap(data))))
```
Comment (olegmoz) [inline src/test/java/com/artipie/asto/s3/S3StorageTest.java:122]: @Sammers21 `OneOffPublisher` usage becomes pointless here, see above

**Comment by Vatavuk** [comment]
@g4s8 please assign this PR to @olegmoz

**Comment by g4s8** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/artipie/asto/issues/207#issuecomment-648103985))

@g4s8 This is what I know about this job in CT2E6TK9B, as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#207](https://github.com/artipie/asto/issues/207) is [in scope](http://www.zerocracy.com/policy.html#14) for 1days
  * The role is `REV`
  * The job is assigned to @olegmoz[/z](https://www.0crat.com/u/olegmoz) for [1days](http://www.zerocracy.com/policy.html#8)
  * There is no monetary reward attached, it's a [free](http://www.zerocracy.com/policy.html#2) job
  * The job doesn't have any [impediments](http://www.zerocracy.com/policy.html#9)
  * The [budget](http://www.zerocracy.com/policy.html#4) is 15 minutes/points
  * These users are banned and won't be assigned:
    * @sammers21[/z](https://www.0crat.com/u/sammers21): This user reported the ticket
    * @vatavuk[/z](https://www.0crat.com/u/vatavuk): User was resigned from the ticket
  * Job [footprint](https://www.0crat.com/footprint/CT2E6TK9B?q=%7Bjob%3A%27gh%3Aartipie%2Fasto%23207%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by 0crat** [comment]
@olegmoz[/z](https://www.0crat.com/u/olegmoz) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by Sammers21** [comment]
@g4s8, can we merge?

**Comment by 0crat** [comment]
Job [#207](https://github.com/artipie/asto/issues/207) is not in the agenda of @olegmoz[/z](https://www.0crat.com/u/olegmoz), can't set title

**Comment by 0crat** [comment]
Job [#207](https://github.com/artipie/asto/issues/207) is not in the agenda of @olegmoz[/z](https://www.0crat.com/u/olegmoz), can't set title

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @olegmoz[/z](https://www.0crat.com/u/olegmoz), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by sereshqua** [comment]
@Sammers21 @olegmoz please make sure you start all you comments with the names of the users they are referred to, see:
[1](https://github.com/artipie/asto/pull/207#issuecomment-647391970), [2](https://github.com/artipie/asto/pull/207#issuecomment-647412071)

**Comment by WhimsySoftware** [comment]
@g4s8 there is definitely something wrong with GitHub repositories in here, I am not logged to my company's acc and I can write comments :)

**Pull request merged**